### PR TITLE
[IMP] mail,portal,test_{mail_full|website}:add language support for portal link

### PR DESCRIPTION
**Pull request by pyduf**
Before this change, the portal link in the email always redirected to a page in the default language (for example while sending a quotation). Now, the portal link redirect to a page in the customer language if it is available.

We add methods in ir_http in portal to easily localize URL so that it can be used in other modules.

Technical notes: when "website" is installed, the portal language is restrained to the website languages. So having a partner with a given language doesn't guarantee that that language is available in the portal. That's why we check the availability of the language before prepending the language to the URL (otherwise it leads to a 404). For that, we call "_get_frontend" with request.is_frontend to True.

Tests have been added in test_mail_full to test the portal behavior alone and in test_website to test the portal behavior with the website installed. This is necessary because installing website restraint the available language of portal to the one of the main website.

In test_mail_full, we patch the method _get_frontend of res_lang to the "http_routing" behavior (i.e.: call _get_frontend with web_force_installed_langs=True context variable). The goal is to ensure that _get_frontend is not limited to the language of the main website when website module is installed and keep the behavior of the portal module without website module regarding the language. Indeed, when installing the module website, some language that were available in the portal become unavailable if not installed on the main website.

Task-2716028

### CODE REVIEW

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Interesting PR, solution is quite short.
I'm interested to know the impact of the change, I'm guessing it add queries? Do you know how many?

Also, seems like we have some test of this route in "test_redirect_to_records", would be interesting to test your use case(s) also, as a bit tricky.

Thanks & cheers!

**Review by pyduf — COMMENTED** [review]
Hello @awa-odoo, 

Thanks for the review. 
Yes indeed, the test were missing. One problem with the test is that the behavior change when installing modules (website override the method get_available of res_lang to restrict to language of the website, which is not the case when website is not present even if portal is not directly related to website). So I have added test in test_mail_full to test portal behavior and in test_website to test it with website (I have commented that in the commit). In the test_mail_full, I patch the get_available method to be sure the language is not restrained by the website that override that method as I have no influence on the website configured language in test_mail_full (not aware of the website module). I have tried to test it with multiwebsite but failed to (forcing the website with MockRequest was not working).
For the added queries, I counted 7 added queries but it is confined in the “_redirect_to_record” method which is used to redirect the user from a link “shared” with an access_token for example in an email. So it has a very narrow impact because it adds query only when the user click on such link, has no access to the record and a token is provided in the URL to give him access. What do you think ?

File: addons/test_mail_full/tests/test_portal.py [hunk 1]
```diff
@@ -278,6 +292,22 @@ def test_employee_access(self):
...
 
+    @users('employee')
+    def test_employee_access_no_language_redirection(self):
+        """Check that the employee is redirected to the record page whatever his language."""
+        # patch: see test_customer_access_not_logged_language_redirection
+        with patch.object(type(self.env['res.lang']), 'get_available',
+                          lambda self: self.with_context(active_test=False).search([]).get_sorted()):
+            self.authenticate(self.env.user.login, self.env.user.login)
+            ResLang_sudo = self.env['res.lang'].sudo()
+            for lang_code, __ in (('fr_FR', '/fr'), ('en_US', ''), ('es_ES', '/es')):
```
Comment (awa-odoo) [inline addons/test_mail_full/tests/test_portal.py:303]: If the prefix is not used, why not just loop on some language codes?

File: addons/test_website/controllers/portal.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+from odoo import http
+from odoo.http import request
+
+
+class PortalTest(http.Controller):
+    """Implements some test portal routes (ex.: for viewing a record)."""
+
+    @http.route('/my/website_test_portal/<int:res_id>', type='http', auth='public', methods=['GET'], website=True)
+    def test_portal_record_view(self, res_id, access_token=None, **kwargs):
+        return request.make_response(f'Record view of test_portal {res_id} ({access_token}, {kwargs})')
```
Comment (awa-odoo) [inline addons/test_website/controllers/portal.py:14]: I don't really understand the purpose of this route?

I get that you will use this syntax in your tests, but do you actually need to define the route? It does not look like you "hit" it.

File: addons/test_website/models/model.py [hunk 3]
```diff
@@ -31,3 +31,23 @@ def _search_get_detail(self, website, order, options):
...
+    _description = 'Chatter Model for Portal'
+    _name = 'website.mail.test.portal'
+    _inherit = [
+        'portal.mixin',
+        'mail.thread',
+    ]
+
+    name = fields.Char('Name')
+    partner_id = fields.Many2one('res.partner', 'Customer')
+    user_id = fields.Many2one(comodel_name='res.users', string="Salesperson")
```
Comment (awa-odoo) [inline addons/test_website/models/model.py:48]: Would leave out "comodel_name", it's not there for the partner_id field anyway.

(I think it's copied as is but we can adapt a bit ^^)

File: addons/test_website/tests/test_redirect.py [hunk 4]
```diff
@@ -231,3 +231,56 @@ def test_05_redirect_404_notfound_record(self):
...
+
+    def test_customer_access_not_logged_language_redirection(self):
+        """Check that the customer is redirected to a page in his language."""
+        ResLang = self.env['res.lang']
+        lang_code = 'en_US'
+        ResLang._activate_lang(lang_code) or ResLang._create_lang(lang_code)
+        self.customer.lang = lang_code
+        res = self.url_open(self.record_access_url)
+        self.assertNotIn(f'/en/my/website_test_portal/{self.record_portal.id}', res.url)
+        self.assertIn(f'/my/website_test_portal/{self.record_portal.id}', res.url)
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:273]: Can't we add this as a lang_code / url_prefix case? So that it fits in your loop?

File: addons/test_website/tests/test_redirect.py [hunk 5]
```diff
@@ -231,3 +231,56 @@ def test_05_redirect_404_notfound_record(self):
...
+
+        for lang_code, url_prefix in (('fr_FR', '/fr'), ('es_ES', '/es')):
+            lang = ResLang._activate_lang(lang_code) or ResLang._create_lang(lang_code)
+            self.customer.lang = lang_code
+
+            # Language not enabled in the current website: no redirection to customer language to avoid 404
+            self.website.language_ids -= lang
+            res = self.url_open(self.record_access_url)
+            self.assertNotIn(f'{url_prefix}/my/website_test_portal/{self.record_portal.id}', res.url)
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:282]: Would still check that we have somewhat of an url inside there, like you're doing above for the en_US use case?

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Thank you for the update and detailed explanation.

About performance and narrow impact, I'm not sure it's that narrow, as it would add those 7 queries every time a customer clicks to "see his invoice", correct? (Which happens a lot I'm guessing).

It depends a bit on the queries ran, if it's just 7 "select * from model where id = x", then it should be fine. The cache may also help.

About the tests, that looks good to me, maybe the reviewer will have more to say on that front.

Added a few polishing comments while re-reading.

Cheers!

**Review by pyduf — COMMENTED** [review]
Hello @awa-odoo, 

Thanks for the review.
Actually, it was not easy to analyze because the queries span multiple http request and then multiple transaction. But with website installed, there is finally 2 additional queries for the code added and yet 3 additional queries when the redirection with the language code in the URL occurs. So it is 5 additional queries. 
The 2 added queries are very simple and indexed: 
–	table res_groups_users_rel queried with UID
–	res.partner queried by id
The 3 other additional queries that happens outside the code because of the added language in the URL are also based on indexes (without the language added in the URL, there is one query instead of 4 with the language in the URL but that will happen to any URL having the language in the URL):
–	res_lang queried by id (2 times, different id)
–	res_lang queried by code (index)
Summary when language should be integrated
–	Without the added code, we have 21 queries + 1 for the redirect
–	With the added code, we have 23 queries + 4 for the redirect (because of the language presence)
Summary when language shouldn’t be integrated (language not installed):
–	Without the added code, we have 21 queries + 1 for the redirect
–	With the added code, we have 23 queries + 1 for the redirect
I have added test with query count that demonstrate the 23 + 4 and the 23 + 1 queries in the test (the 2 lines above with the code added).

For the case of portal installed without website, I haven’t analyzed everything in details but have added the query count test. The added code adds 4 queries but simple and indexed (except condition in the second but it is for default value and already 3 indexed field in the condition):
–	res_partner by id
–	get_default_lang for res.partner (query on field, user_id, company_id all indexed and condition not indexed)
–	ir_default by id 
–	res_lang by id
Strangely there are more queries when using the default language. I haven’t digged in details but it is probably due to the fact that “url_for” is counter productive this time because it adds the prefix ‘/en’ to the URL which is removed after the redirection.

In summary, with only portal installed (not website), we have 
–	15 queries all time when not installed 
–	20 when redirecting to a language different than the default one
–	28 when redirecting to a language different than the default one (due probably to the redirection that update the URL to remove the “/en”)
But in the added code, we only have 4 simple added queries as explained above. I haven’t added the profile tests because it will have other query count when website is installed.

The number are when installing portal and/or website. When other modules are installed, there are 2 added queries (observed on the runbot).

Do you think those added queries are worth the feature ?

File: addons/test_website/tests/test_redirect.py [hunk 6]
```diff
@@ -231,3 +232,91 @@ def test_05_redirect_404_notfound_record(self):
...
+            request.env['res.lang'].with_context(website_id=self.website.id).sudo().get_available()
+            request.env['res.lang'].sudo().get_available()
+
+    def setup_website_customer_lang(self, lang_code):
+        """ Activate language for the website and set it to the customer. """
+        ResLang = self.env['res.lang']
+        lang = ResLang._activate_lang(lang_code) or ResLang._create_lang(lang_code)
+        self.website.language_ids += lang
+        self.customer.lang = lang_code
+        return lang
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:280]: If those are utility methods, would make them private and move them to the and of the file.

File: addons/test_website/tests/test_redirect.py [hunk 7]
```diff
@@ -231,3 +232,91 @@ def test_05_redirect_404_notfound_record(self):
...
+        ):
+            if group_name == 'portal_customer' and group_func(cls.customer):
+                cls.record_access_url = group_data['button_access']['url']
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        cls.website = cls.env.ref('website.default_website')
+
+    def preload_res_lang_get_available(self):
+        """ Preload get_available language for default website and no website (see url_lang called by url_for) """
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:268]: That is to "cache" those queries to remove them from the query counters below then?

Would be more specific in the docstring if so.

File: addons/test_website/tests/test_redirect.py [hunk 8]
```diff
@@ -231,3 +232,91 @@ def test_05_redirect_404_notfound_record(self):
...
+    def setup_website_customer_lang(self, lang_code):
+        """ Activate language for the website and set it to the customer. """
+        ResLang = self.env['res.lang']
+        lang = ResLang._activate_lang(lang_code) or ResLang._create_lang(lang_code)
+        self.website.language_ids += lang
+        self.customer.lang = lang_code
+        return lang
+
+    @mute_logger('odoo.addons.base.models.ir_model')
+    def test_customer_access_not_logged_language_redirection(self):
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:283]: Isn't this one redundant with your other "test_customer_access_not_logged_language_redirection" ?

File: addons/test_website/tests/test_redirect.py [hunk 9]
```diff
@@ -231,3 +232,91 @@ def test_05_redirect_404_notfound_record(self):
         r = self.url_open(url_rec1, allow_redirects=True)
         self.assertEqual(r.status_code, 200)
         self.assertTrue(r.url.endswith(redirect.url_to))
+
+
+@tagged('-at_install', 'post_install')
+class TestRedirectPortal(HttpCase):
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:238]: If we end up removing "test_customer_access_not_logged_language_redirection", would rename to "TestPortalRedirectionPerformance", as this one seems more performance-oriented.

File: addons/test_website/tests/test_redirect.py [hunk 10]
```diff
@@ -231,3 +232,91 @@ def test_05_redirect_404_notfound_record(self):
...
+            self.assertNotIn(f'{expected_url_prefix}/my/website_test_portal/{self.record_portal.id}', res.url)
+            self.assertIn(f'/my/website_test_portal/{self.record_portal.id}', res.url)
+            # Language enabled in the current website: redirection to customer language
+            self.website.language_ids += lang
+            res = self.url_open(self.record_access_url)
+            self.assertIn(f'{expected_url_prefix}/my/website_test_portal/{self.record_portal.id}', res.url)
+
+    @mute_logger('odoo.addons.base.models.ir_model')
+    def test_profile_language_redirection(self):
+        self.setup_website_customer_lang('es_ES')
```
Comment (awa-odoo) [inline addons/test_website/tests/test_redirect.py:304]: I would add some docstring to each testing case to explain the query differences.

Can be a subset of your big explanation of the PR :smile:

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Thank you for the very detailed explanation.
The feature still seems worthwhile as we have only a few added queries and those are fast (and probably cached) ones.

I added a few extra comments.
Once handled, feel free to green the task so that it can move forward in the pipre during my vacation.

Thanks again for the followup and cheers!

**Review by pyduf — COMMENTED** [review]
Hello @awa-odoo , 

Thanks for the review and your useful advice.  I have done the changes.
